File: content/examples/other/leetcode/problems/coin-change/example-1/solution.py
def coin_change_top_down(coins, amount):
    memo = {}

    def dfs(remaining):
        if remaining == 0:
            return 0
        if remaining < 0:
            return float("inf")
        if remaining in memo:
            return memo[remaining]
        best = float("inf")
        for c in coins:
            sub = dfs(remaining - c)
            if sub != float("inf"):
                best = min(best, sub + 1)
        memo[remaining] = best
        return best

    res = dfs(amount)
    return -1 if res == float("inf") else res


if __name__ == "__main__":
    print(coin_change_top_down([1, 2, 5], 11))

File: content/examples/other/leetcode/problems/coin-change/example-1/solution.js
function coinChangeTopDown(coins, amount) {
  const memo = new Map();
  function dfs(remaining) {
    if (remaining === 0) return 0;
    if (remaining < 0) return Infinity;
    if (memo.has(remaining)) return memo.get(remaining);
    let best = Infinity;
    for (const c of coins) {
      const sub = dfs(remaining - c);
      if (sub !== Infinity) best = Math.min(best, sub + 1);
    }
    memo.set(remaining, best);
    return best;
  }

  const res = dfs(amount);
  return res === Infinity ? -1 : res;
}

if (require.main === module) {
  console.log(coinChangeTopDown([1, 2, 5], 11));
}

module.exports = { coinChangeTopDown };
